### PR TITLE
Linting and code formatting

### PR DESCRIPTION
**Pull request by katoss**
Hi everyone, 

in this PR/on this branch I'm trying to collect commits related to general code formatting and linting.

These are the issues I've addressed or plan to address (I added names according to who gave me feedback on this, but of course feel free to review any of them!):

1. **Linting** 
- @Robaina, @Batalex, @khynder
- 4e1a75b5501398bf8d03b552a8e673185a3cab44, 91482b17c0f261d4b1300b2ef8e546f536f2e239: Added black to poetry and formatted the analysis file with it
- **Open Question:** I am wondering now where to best integrate black, I read that it is possible as pre-commit hook and in the CI (as suggested by @Robaina ). Maybe even both? Then there would be a local check before committing, and also a later check in case some contributor does not use the pre-commit hook. What do you think?

2. **Type hints**
- 0fe257e0b1f95b7ad0a8e506f604a9386015baae: Added type hints (@Robaina )

3. **Specific formatting issues:**
- ca2c2dec908d01afab52781a7db4e660726ce1a8: Moved "\_\_all_\_\" below imports (@Batalex)
- 263df2fa5fa90b303a3e9de407cc44faffa4cf35, ffcd08baed4dc825855c641f0a3ecabebf7b5f9a: Replaced concats with format strings (@Robaina , @Batalex)

Best,
@katoss

### CODE REVIEW

**Comment by Batalex** [comment]
You can take a look at https://pre-commit.ci/ so that you can integrate black in your CI

File: pyproject.toml [hunk 1]
```diff
@@ -12,6 +12,7 @@ pandas = "^1.5.0"
 numpy = "^1.20.3"
 matplotlib = "^3.3.2"
 scipy = "^1.7.1"
+black = "^23.3.0"
```
Comment (Batalex) [inline pyproject.toml]: `black` should be a dev dependency, as it is not needed for runtime

File: src/cardsort/analysis.py [hunk 2]
```diff
@@ -1,30 +1,35 @@
...
-from scipy.spatial.distance import squareform
-
 
-def _get_distance_matrix_for_user(df_user):
+def _get_distance_matrix_for_user(df_user: pd.DataFrame) -> np.ndarray:
     n = len(df_user)
 
     X = np.zeros((n, n))
     for i in range(n):
         for j in range(n):
```
Comment (Robaina) [inline src/cardsort/analysis.py]: You would save a bunch of computations if you use here:

```python
for i in range(n):
    for j in range(i, n):
```
so you only iterate over the upper triangle of your symmetric matrix. You are already copying the entries to the lower triangle below:

```python
X[i, j] = X[j, i] = 0 if cat1 == cat2 else 1
```

**Review by Robaina — COMMENTED** [review]
Nice use of typing! much clearer now

File: pyproject.toml (hunk 1 above)
Comment (katoss) [reply pyproject.toml]: oh sure my bad, fixed it. Thank you!

File: src/cardsort/analysis.py (hunk 2 above)
Comment (katoss) [reply src/cardsort/analysis.py]: Thank you, fixed it

**Comment by katoss** [comment]
> You can take a look at https://pre-commit.ci/ so that you can integrate black in your CI

@Batalex I have not entirely understood this service yet. I add black to the pre-commit file (which I haven't created it), and then this service will automatically perform the precommit as part of the CI?

**Comment by Batalex** [comment]
> > You can take a look at https://pre-commit.ci/ so that you can integrate black in your CI
> 
> @Batalex I have not entirely understood this service yet. I add black to the pre-commit file (which I haven't created it), and then this service will automatically perform the precommit as part of the CI?

Yes, that is the idea. I have yet to use this service myself, so this is not by any means required for our review process.

**Comment by katoss** [comment]
> Yes, that is the idea. I have yet to use this service myself, so this is not by any means required for our review process.

I tried it and it seems to work :) I will convert this draft to a PR, I think we are good to go with this one, too

File: tests/test_analysis.py [hunk 3]
```diff
@@ -6,19 +6,64 @@
 
 
 def test_get_distance_matrix():
-    expected = [2., 1., 5., 5., 5., 4., 5., 5., 5., 1., 5., 5., 
-                5., 5., 5., 5., 4., 5., 5., 5., 5., 5., 5., 5.,
-                2., 1., 4., 3., 3., 5., 3., 2., 1., 1., 5., 3., 
-                3., 3., 5., 2., 2., 5., 0., 4., 4.]
+    expected = [
```
Comment (Batalex) [inline tests/test_analysis.py]: If you'd like, you could use comments to disable black for these lines:

```py
# fmt: off
expected = [...]

# fmt: on

```

**Review by Batalex — APPROVED** [review]
I noticed a few typos in the code (`treshold` instead of `threshold`). It's up to you if you want to fix it in this PR.
Otherwise, LGTM ✔︎

**Review by Robaina — APPROVED** [review]

File: tests/test_analysis.py (hunk 3 above)
Comment (katoss) [reply tests/test_analysis.py]: oh yes, great, I already wondered if this is really the intended behavior in this case, listing one value in each line. I didn't know that black could be disabled for specific lines, will do

**Pull request merged**